### ISSUE TITLE
Since Powershell 7.4, [PSCustomObject] typecast results in a null value. 

### ISSUE THREAD

**Issue by dam-pav** [PowerShell/PowerShell#20934]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

It would appear [PSCustomObject] is broken. I realized this because a lot of code stopped working after upgrading to 7.4. In 5.1 same code still works. Consider this test:

```
Clear-Host

Write-Host ''
Write-Host 'Hash:' -ForegroundColor Green
$newHash = @{
    AppId        = '12341234'
    Version      = '23452345'
    Name         = 'whatever'
    Publisher    = 'default'
}
$newHash


Write-Host ''
Write-Host 'Object with [PSCustomObject]:' -ForegroundColor Green
$newObject = [PSCustomObject]$newHash
$newObject


Write-Host ''
Write-Host 'Object with New-Object:' -ForegroundColor Green
$newObject = New-Object -TypeName psobject -Property $newHash
$newObject


Write-Host 'Done.' -ForegroundColor Green

```

### Expected behavior

```console
Hash:

Name                           Value
----                           -----
AppId                          12341234
Publisher                      default
Name                           whatever
Version                        23452345

Object with [PSCustomObject]:

AppId     : 12341234
Publisher : default
Name      : whatever
Version   : 23452345


Object with New-Object:

AppId     : 12341234
Publisher : default
Name      : whatever
Version   : 23452345

Done.
```

### Actual behavior

```console
Hash:

Name                           Value
----                           -----
Version                        23452345
Publisher                      default
AppId                          12341234
Name                           whatever

Object with [PSCustomObject]:

Object with New-Object:

Version   : 23452345
Publisher : default
AppId     : 12341234
Name      : whatever

Done.
```

### Environment data

```powershell
Both environments:
PS C:\Projects\__KB\Docker> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.22000.2600
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22000.2600
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1


PS C:\Projects\__KB\Docker> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

Results in 5.1:
<img width="254" alt="image" src="https://github.com/PowerShell/PowerShell/assets/46426230/43112558-93e6-447b-866b-cd490c474429">

Results in 7.4:
<img width="234" alt="image" src="https://github.com/PowerShell/PowerShell/assets/46426230/c8975f91-a60d-4567-926b-10fb76bc9184">

**Comment by mklement0** [comment]
This sounds like a duplicate of:

* #20768

The latter is already closed because the problem was fixed in #20913.
Judging by the [BackPort-7.4.x-Consider](https://github.com/PowerShell/PowerShell/labels/BackPort-7.4.x-Consider) label on the latter, the plan is to release the fix in v7.4.1

**Comment by StevenBucher98** [comment]
Hmm I am not able to reproduce on 7.4, I copied the code you had and this is the output:

![image](https://github.com/PowerShell/PowerShell/assets/20605456/1991d35d-2bd1-4308-8f89-ca6e5e28511b)

Any additional insight about your environment that you think may be causing this? Perhaps it will get fixed in 7.4.1 like @mklement0 says.

**Comment by dam-pav** [comment]
I verified @mklement0 's claim and yes, this definitely is related to $ExecutionContext.SessionState.LanguageMode. Setting this to 'FullLanguage' does restore the casting results, but only within a very limited context (local to the function level from what I can tell). As the MS documentation says (can provide link), setting this explicitly is not the intended method. 

Unfortunately I was not able, so far, to understand what the actually intended method was, ie. to set this on a session level. My use case is scripting development environments, from VSCode, using imported PS modules. I can adapt my own scripts sure, however the modules are littered with typecasting. Obviously and rightfully so. I'd have to learn more about WDAC, but so far I understand that this is a security framework for policies set on an organizational level. 

I was looking for any way to control this on a session level since these sessions are started from VSCode/PS extension. So far no luck. I found references, but none of that works (can provide links if interested).

So what we are left with is corporate PCs unable to do any corporate work. 

I don't have links ready because this is not the station where I did the research.

Bottom line, I really do hope LanguageMode is affecting [PSCustomObject] by mistake and this is actually going away with 7.4.1. He says it _may_. :-|

**Comment by mklement0** [comment]
> He says it _may_

I certainly hope that it _will_, but I neither have inside knowledge nor any say in the matter.

The _may_  (a qualification I added later) is purely based on the intended use of the `BackPort-7.4.x-Consider` label: as I later realized, it means just that: the fix is being _considered_ for backporting, but backporting hasn't yet been _approved_ (see https://github.com/PowerShell/PowerShell/issues/20895#issuecomment-1850723132 for related labels and what I think is the intended workflow)

**Comment by mklement0** [comment]
@dam-pav, as for a _stopgap_, based on your environment description: perhaps the following works: enable profile loading in the PowerShell extension, open the profile for editing (`psedit $PROFILE`)  and place `$ExecutionContext.SessionState.LanguageMode = 'FullLanguage'` at the top.

**Comment by dam-pav** [comment]
@mklement0, thanks for the idea, unfortunately it doesn't work.

Maybe of some interest, none of the profiles are present by any default. That includes the 'Microsoft.VSCode_profile.ps1', the special VSCode extension profile. (https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_profiles?view=powershell-7.4)



Ok then, I tried creating scripts both for "All Users, All Hosts" and for "Current User, VS Code". I get this:
![image](https://github.com/PowerShell/PowerShell/assets/46426230/39616bb9-45f6-45ee-a803-147d3c120c8c)

while loading VSCode's PS session.

Usually you will get nothing, no messages. As my previous examples show. Additional profile entries appear to have no impact, until this. The exception messages are new, and are somehow triggered because I set these default policies:
![image](https://github.com/PowerShell/PowerShell/assets/46426230/c8b5f635-5ff6-4846-8cbc-6fff46144ad3)

It shows at least that the profile entries indeed do not have an impact. With or without them, the test script has the same output (now with additional exception message):
![image](https://github.com/PowerShell/PowerShell/assets/46426230/6693334c-dddb-4c06-bfbf-9c008715648d)

And another puzzling fact, if I run this as administrator (the whole VSCode session) the exception messages are gone again, but cast to [PSCustomObject] still returns null.

**Comment by dam-pav** [comment]
So I decided I buried enough time into this and went back to using PS 5.1. But, I forgot to remove the default security policies. So get this:
![image](https://github.com/PowerShell/PowerShell/assets/46426230/1b743e25-e60c-457d-bc5c-720952f7c440)

The cast supposedly fails, yet the object gets created. Excuse me, but _wot_?

**Comment by mklement0** [comment]
@dam-pav, my knowledge of PowerShell's security features in general is superficial, but perhaps the following pointers help:

* From what I understand, WDAC, which the linked issue is related to, effectively superseded AppLocker, which the `secpol.msc` screenshot you're showing is related to.

* The linked issue is about a newly introduced WDAC _audit-only_ variant of constrained-language mode, which - whether by design or not, I do not know - still _reports_ as if it _were_ the actual constrained-language mode via `$ExecutionContext.SessionState.LanguageMode` 

* It is only in the context of this _audit-only_ mode that executing `$ExecutionContext.SessionState.LanguageMode = 'FullLanguage'` is even _possible_ - whether it _should_ be possible, I don't know.

* The error message you're seeing - `Cannot set property. Property setting is supported only on core types in this language mode.` - implies that you're in _actual_ constrained-language mode, which - by design - _prevents_ switching back to `FullLanguage` mode (or any other mode).

As an aside: 
 * The inconsistency that  in actual constrained-language mode you can create `[pscustomobject]` instances with `New-Object`, but not with `[pscustomobject]` _literals_ is the subject of #20767.

**Comment by dam-pav** [comment]
@mklement0, thanks again. The inconsistency of [PSCustomObject] vs. New-Object stands out, agreed. However, it would now appear that this inconsistency is not limited to PS7.4. 

The questions keep opening and don't have neither the expertise nor the time to deal with all that. They seem to be worthy of investigation however and it looks like they already are being investigated, for the most part. The inability to use any of that is mostly mine and it would appear there is nothing much else I can do but to wait for a packaged solution (hopefully in the form of release 7.4.1). That much is clear, to me. 

Thanks again for all your engagement and the time it took.
